src/Microsoft.AspNet.Routing/Template/TemplateRoute.cs

```diff
@@ -203,19 +203,23 @@ public virtual string GetVirtualPath(VirtualPathContext context)
             // Validate that the target can accept these values.
             var childContext = CreateChildVirtualPathContext(context, values.AcceptedValues);
 
-            var path = _target.GetVirtualPath(childContext);
-            if (path != null)
+            var pathData = _target.GetVirtualPath(childContext);
+            if (pathData != null)
             {
                 // If the target generates a value then that can short circuit.
-                return path;
+                return pathData;
             }
 
             // If we can produce a value go ahead and do it, the caller can check context.IsBound
             // to see if the values were validated.
-            path = _binder.BindValues(values.AcceptedValues);
+            pathData = new VirtualPathData(
+                router: this,
+                virtualPath:_binder.BindValues(values.AcceptedValues),
+                dataToken: _dataTokens ?? RouteValueDictionary.Empty);
```

Comment:
System.Web.Routing always copied the data tokens, so I think we should do the same. The VirtualPathData is a snapshot in time so we don't want to point to the original reference.